### PR TITLE
[add] Image source headers handling

### PR DESCRIPTION
**Pull request by kidroca**
Upstream PR: https://github.com/necolas/react-native-web/pull/2442

## Details 

Extend `ImageLoader` functionality to be able to work with image sources containing headers

We preserve the existing strategy that works with `image.src` for cases where  `source` 
is just an `uri` with no headers

When `source` contain headers we make a `fetch` request and then render a local url for the
downloaded blob (`URL.createObjectURL`)

## Fixed Issues

$ https://github.com/Expensify/App/issues/12603

## Test Strategy

1. Verify existing Image functionality has no regressions
    - build web and examples: ` npm run dev -w react-native-web` and `npm run dev -w react-native-web-examples`
    - open the examples page and go to `Image`: http://localhost:3000/image
    - see images are loading
    - take a screenshot and do the same from the `master` branch. You can switch back and forth and verify the image are loading the same way

2. Verify Images with headers can be loaded
    - build web and examples: ` npm run dev -w react-native-web` and `npm run dev -w react-native-web-examples`
    - open the examples page and go to `Image`: http://localhost:3000/image
    - modify `sourceWithHeaders` here `packages/react-native-web-examples/pages/image/index.js` and try to load images from a server that expects a GET request with a header
    - verify the image is loading on the examples page (near the bottom, labeled: "With Headers")

### CODE REVIEW

File: packages/react-native-web/src/exports/Image/index.js
Comment (kidroca) [inline packages/react-native-web/src/exports/Image/index.js]: I've tried to write some unit tests covering the new functionality, but because of this logic a few mocks need to be setup and make the PR bigger than it has to be 

I think it might be better to move this logic to `ImageLoader.loadUsingHeaders` so in unit tests we can stub that method and ensure it gets called when we intent to 

It's easy to manually verify the `fetch` logic works and downloads content 
Once that's clear it's easy to just verify `loadUsingHeaders` is called correctly with expected input

---

I've made a similar comment on the main repo: https://github.com/necolas/react-native-web/pull/2442/files#r1072080652

--- 

What do you think, should we refactor / write some tests now or let the mainstream repo request those changes?

**Comment by kidroca** [comment]
In need of some Review feedback here

cc @Beamanator @marcaaron @tgolen

File: packages/react-native-web/src/modules/ImageLoader/index.js
Comment (Beamanator) [inline packages/react-native-web/src/modules/ImageLoader/index.js:129]: Mind dropping a comment here why this is needed?

File: packages/react-native-web/src/exports/ImageBackground/index.js
Comment (Beamanator) [inline packages/react-native-web/src/exports/ImageBackground/index.js:19]: It looks like you're trying to use "Exact object types" from Flow: https://flow.org/en/docs/types/objects/#toc-exact-object-types - I feel like since we're trying to keep this PR as lean as possible I'd prefer removing this, let me know if you have a different opinion

**Comment by Beamanator** [comment]
Sorry I didn't have a chance to get to this today, @marcaaron do you mind giving this a first pass if you have time today? I will definitely review tomorrow no matter what 👍

**Comment by marcaaron** [comment]
Chatted 1:1 with @Beamanator. I'd love to see a good faith effort on this one before chiming in. It sounds like there are some higher priority things on Alex's plate that kept him from getting to this today. Taking that information... I've determined that my review is not urgently needed here and Alex has agreed to give a first review (I think).

File: packages/react-native-web/src/modules/ImageLoader/index.js
Comment (Beamanator) [inline packages/react-native-web/src/modules/ImageLoader/index.js:132]: Won't `image.src` always be `''`, and `image.naturalWidth` and `image.naturalHeight` will always be `0` at this point since we just created a brand new `window.Image`?

File: packages/react-native-web/src/exports/Image/types.js
Comment (Beamanator) [inline packages/react-native-web/src/exports/Image/types.js:106]: Is there a reason you're adding this ([docs ref](https://flow.org/en/docs/types/utilities/#toc-exact)) other than personal pref? I think we're preferring to make this as lean as possible, no?

File: packages/react-native-web/src/exports/Image/index.js
Comment (Beamanator) [inline packages/react-native-web/src/exports/Image/index.js:163]: If `a.headers` has properties in a different order than `b.headers`, then this will return true even though they could be the exact same, no? That said, in our case we only want to add 1 header at a time so that's not an issue - should we add a comment about that?

File: packages/react-native-web/src/modules/ImageLoader/index.js
Comment (Beamanator) [reply packages/react-native-web/src/modules/ImageLoader/index.js:132]: Not sure if it's useful to add a comment about that just to be 100% clear

File: packages/react-native-web/src/exports/Image/index.js
Comment (Beamanator) [inline packages/react-native-web/src/exports/Image/index.js]:
```suggestion
    // Capture a cleanup function for the current request.
    // The reason for using a Ref is to avoid making this function a dependency
    // because the change of a dependency would otherwise re-trigger a hook.
```

File: packages/react-native-web/src/exports/Image/index.js
Comment (Beamanator) [inline packages/react-native-web/src/exports/Image/index.js]: So if I understand this correctly, we're basically storing a function here that will be used to clean up an existing image fetch request - and we'll only actually call that function if `nextSource` changes, yeah? Nice 👍

File: packages/react-native-web/src/exports/Image/index.js
Comment (Beamanator) [inline packages/react-native-web/src/exports/Image/index.js]: Make make a note that this doesn't load images with headers passed in `defaultSource` so that's perfectly clear

File: packages/react-native-web/src/exports/Image/index.js
Comment (Beamanator) [reply packages/react-native-web/src/exports/Image/index.js]: Hmm since this logic is the real meat & potatoes of the `ImageWithHeaders` component, if we move it to `ImageLoader.loadUsingHeaders` I assume we could get rid of this `ImageWithHeaders` component, and basically check if there's headers passed in the props - if yes, we call `ImageLoader.loadUsingHeaders` instead of `ImageLoader.load` here?

https://github.com/Expensify/react-native-web/blob/bd1f7b887ad690ef926f0c3d9f97c1f625d4bd55/packages/react-native-web/src/exports/Image/index.js#L269-L294

If that's what you're thinking, I honestly do think that logic is cleaner, even without needing to write tests for Expensify's case so I'd say the refactor sounds like a nice idea

**Review by Beamanator — COMMENTED** [review]
As a whole I'd say these changes were much easy to follow, thanks for this PR @kidroca 👍 

As I responded to your comment, I do think the refactor (moving `ImageWithHeaders`'s loading logic into `ImageLoader`) could make this even more understandable, as long as that eliminates the need for the `ImageWithHeaders` component. Let me know what you think about my other comments below 👍

File: packages/react-native-web/src/exports/ImageBackground/index.js
Comment (kidroca) [reply packages/react-native-web/src/exports/ImageBackground/index.js:19]: The change is in order to address a flow error (Sorry, I should have mentioned why I had to change types) 

<details>
<summary>Flow Error</summary>

```
Cannot spread props because inexact props [1] is incompatible with exact `ImageProps` [2]. [incompatible-exact]

   packages/react-native-web/src/exports/ImageBackground/index.js:46:7
              v-----
    46|       <Image
    47|         {...rest}
    48|         ref={imageRef}
    49|         style={[
    50|           {
    51|             // Temporary Workaround:
    52|             // Current (imperfect yet) implementation of <Image> overwrites width and height styles
    53|             // (which is not quite correct), and these styles conflict with explicitly set styles
    54|             // of <ImageBackground> and with our internal layout model here.
    55|             // So, we have to proxy/reapply these styles explicitly for actual <Image> component.
    56|             // This workaround should be removed after implementing proper support of
    57|             // intrinsic content size of the <Image>.
    58|             width,
    59|             height,
    60|             zIndex: -1
    61|           },
    62|           StyleSheet.absoluteFill,
    63|           imageStyle
    64|         ]}
    65|       />
              -^ [1]

References:
   packages/react-native-web/src/exports/Image/index.js:173:3
   173|   ImageProps,
          ^^^^^^^^^^ [2]
```
</details>

Usually it's preferable to use exact types for props

And this is the reason why Image props need to be exact 

<details>
<summary>Flow Error</summary>

```
Error ------------------------------------------------------ packages/react-native-web/src/exports/Image/index.js:413:10
> npm run flow && npm run format && npm run lint && npm run unit --runInBand
Cannot spread props because Flow cannot determine a type for props [1]. object literal [2] is inexact, so it may contain
`ref` with a type that conflicts with `ref`'s definition in props [1]. Try making object literal [2] exact.
[cannot-spread-inexact]

   packages/react-native-web/src/exports/Image/index.js:413:10
   413|   return <BaseImage ref={ref} {...propsToPass} />;
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ [1]

References:
   packages/react-native-web/src/exports/Image/index.js:403:23
                              v
   403|   const propsToPass = {
   404|     ...props,
   405|     // Omit `onLoadStart` because we trigger it in the current scope
   406|     onLoadStart: undefined,
   407|     // Until the current component resolves the request (using headers)
   408|     // we skip forwarding the source so the base component doesn't attempt
   409|     // to load the original source
   410|     source: blobUri ? { ...nextSource, uri: blobUri } : undefined
   411|   };
          ^ [2]


Error ------------------------------------------------------ packages/react-native-web/src/exports/Image/index.js:419:12

Cannot spread props because Flow cannot determine a type for props [1]. `ImageProps` [2] is inexact, so it may contain
`ref` with a type that conflicts with `ref`'s definition in props [1]. Try making `ImageProps` [2] exact.
[cannot-spread-inexact]

   packages/react-native-web/src/exports/Image/index.js:419:12
   419|     return <ImageWithHeaders ref={ref} {...props} />;
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ [1]

References:
   packages/react-native-web/src/exports/Image/index.js:173:3
   173|   ImageProps,
          ^^^^^^^^^^ [2]


Error ------------------------------------------------------ packages/react-native-web/src/exports/Image/index.js:422:10

Cannot spread props because Flow cannot determine a type for props [1]. `ImageProps` [2] is inexact, so it may contain
`ref` with a type that conflicts with `ref`'s definition in props [1]. Try making `ImageProps` [2] exact.
[cannot-spread-inexact]

   packages/react-native-web/src/exports/Image/index.js:422:10
   422|   return <BaseImage ref={ref} {...props} />;
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ [1]

References:
   packages/react-native-web/src/exports/Image/index.js:173:3
   173|   ImageProps,
          ^^^^^^^^^^ [2]



Found 3 errors
```
</details>

It was easier to fix the prop type definition than to put $FlowFixMe on a few lines

File: packages/react-native-web/src/exports/Image/index.js
Comment (kidroca) [reply packages/react-native-web/src/exports/Image/index.js:163]: > If `a.headers` has properties in a different order than `b.headers`, then this will return true even though they could be the exact same, no?

Yes, if for some reason headers are swapped the JSON string would be different, though that's quite the edge case, I don't even know how would it happen - if `headers` is build programmatically and starts as a blank `{}`, and then logic adds the headers - it would aways add them in the same order 

If headers are defined inline they always be the same 

Worse case - headers are swapped and this produces a source diff and a new request

---

BTW if you've already made a request (recently) the result gets cached and even if you make the same request with different headers you'd instantly get result from cache 

So it's possible to drop this logic, but it might result in an active request (no cache yet) getting aborted and restarted

File: packages/react-native-web/src/exports/Image/index.js
Comment (kidroca) [reply packages/react-native-web/src/exports/Image/index.js]: Well, the idea is purely to ease mocking but I'll give your suggestion a try

Last PR tried to solve everything inside the same component but that resulted in more logic

What a component like `ImageWithHeaders` gives us is a guarantee that `source` would always be an object with headers

BTW there's feedback on the mainstream PR: https://github.com/necolas/react-native-web/pull/2442#issuecomment-1396183044 that we should write tests and thumbs up for extracting `ImageLoader.loadUsingHeaders`

File: packages/react-native-web/src/exports/Image/types.js
Comment (kidroca) [reply packages/react-native-web/src/exports/Image/types.js:106]: Yeah, I've posted about this in a another comment regarding types: https://github.com/Expensify/react-native-web/pull/11/files#r1081137857

File: packages/react-native-web/src/modules/ImageLoader/index.js
Comment (kidroca) [reply packages/react-native-web/src/modules/ImageLoader/index.js:132]: > Won't `image.src` always be `''`, and `image.naturalWidth` and `image.naturalHeight` will always be `0` at this point since we just created a brand new `window.Image`?

We are inside `image.onload` - the logic here is evaluated later, when the image did load so `src` and dimensions shouldn't be missing

You can see that `Image.getSize` pretty much does the same thing 

https://github.com/Expensify/react-native-web/blob/8f4d952b7966fd6e10831d299bdf76929dac5bc5/packages/react-native-web/src/modules/ImageLoader/index.js#L93-L100

_`callback` is called `onLoad`_

File: packages/react-native-web/src/exports/Image/index.js
Comment (kidroca) [reply packages/react-native-web/src/exports/Image/index.js]: I've tried to remove the `ImageWithHeaders` component and use either `loadUsingHeaders` or `load` functions here: https://github.com/Expensify/react-native-web/compare/master...kidroca:react-native-web:kidroca/feat/image-loader-headers-alt-2?diff=unified

But it results in a similar amount of changes and modifies some of the original logic (and seems harder to review)
- because now the source loading `useEffect` needs to account for objects
- `load` and `loadWithHeaders` need to work in a similar way in order to be interchangeable (so they were modified to return a `request.cancel` function)

We might merge `load` and `loadUsingHeaders` instead of testing which one to run, though this would be similar to the first PR were `load` handled loading images with and without headers

File: packages/react-native-web/src/exports/Image/index.js
Comment (Beamanator) [reply packages/react-native-web/src/exports/Image/index.js:163]: > if `headers` is build programmatically and starts as a blank `{}`, and then logic adds the headers - it would aways add them in the same order 

True true, I'm down to leave it as is since you have good points, I can't think of a way of this to "break" - and "break" (worst case) only means there's an extra request 🤷 

> So it's possible to drop this logic, but it might result in an active request (no cache yet) getting aborted and restarted

Agreed let's keep it for now 👍

File: packages/react-native-web/src/modules/ImageLoader/index.js
Comment (Beamanator) [reply packages/react-native-web/src/modules/ImageLoader/index.js:132]: Aah yepppp that makes sense, thanks for pointing that out 👍

File: packages/react-native-web/src/exports/Image/index.js
Comment (Beamanator) [reply packages/react-native-web/src/exports/Image/index.js]: Thanks for doing that refactor! Personally I prefer the new changes because it keeps all the image loading logic in `ImageLoader` and there's minimal changes to `Image/index.js` - I don't see those new changes too difficult to review (though I am not sure where `lastLoadedSource` gets updated).

I'd say let's move forward with this last refactor, as I think it's pretty straightforward compared to passing / not passing specific props to the BaseImage component required in this PR

Comment (kidroca) [reply packages/react-native-web/src/exports/Image/index.js]: IMO the refactor is more of a proof that there are more "gymnastics" necessary in order to make this work by changing the original logic inside Image component 

Original logic is changed, the cleanup logic is different, `ImageLoader.load` is changed, people, including the mainstream maintainer would have to inspect how these used to work and whether the change is suitable 

IMO the mainstream maintainer already saw the update and is fine with just moving the `fetch` call to `ImageLoader.loadUsingHeaders`. I'm still trying different things, but the most straightforward PR so far seems to be the current one

There's also a big rework planned for the Image and the original loading logic would change, it would probably be best to not write stuff that depend on it (in the [alt branch](https://github.com/Expensify/react-native-web/compare/master...kidroca:react-native-web:kidroca/feat/image-loader-headers-alt-2?diff=unified) `loadUsingHeaders` depends on `load`)

Comment (kidroca) [reply packages/react-native-web/src/exports/Image/index.js]: Let me update the current PR with `loadUsingHeaders` extracted to `ImageLoader` and then we can make one final decision  which set of changes would work best for us

Comment (Beamanator) [reply packages/react-native-web/src/exports/Image/index.js]: > IMO the mainstream maintainer already saw the update and is fine with just moving the fetch call to `ImageLoader.loadUsingHeaders`. I'm still trying different things, but the most straightforward PR so far seems to be the current one

Yeah this is one additional reason I really like this approach, even though there may be some additional changes in the upstream repo needed that we don't need at the moment in this fork 👍 

> Let me update the current PR with `loadUsingHeaders` extracted to `ImageLoader` and then we can make one final decision which set of changes would work best for us

That sounds perfect, thanks so much @kidroca 👍

File: packages/react-native-web/src/exports/ImageBackground/index.js
Comment (Beamanator) [reply packages/react-native-web/src/exports/ImageBackground/index.js:19]: Got it, thanks for the explanation!

File: packages/react-native-web/src/exports/Image/index.js
Comment (kidroca) [inline packages/react-native-web/src/exports/Image/index.js:291]: extracted to `raiseOnErrorEvent` for reuse

File: packages/react-native-web/src/exports/Image/index.js
Comment (kidroca) [inline packages/react-native-web/src/exports/Image/index.js:360]: `ImageLoader.loadUsingHeaders` returns a request with reference to the last loaded source, and a cleanup function. We no longer need to capture `lastLoadedSource` and `cleanup` refs

File: packages/react-native-web/src/exports/Image/index.js
Comment (kidroca) [inline packages/react-native-web/src/exports/Image/index.js:383]: When an error happens we perform the original onError logic, that was moved to `raiseOnErrorEvent`

**Review by kidroca — COMMENTED** [review]
Moved logic to `ImageLoader.loadUsingHeaders`

✅ PR ready for review

And here's an alternative version where we try to use a single Image component instead of creating `BaseImage` and `ImageWithHeaders`: https://github.com/Expensify/react-native-web/pull/13/files

cc @roryabraham

File: packages/react-native-web/src/exports/Image/index.js
Comment (Beamanator) [inline packages/react-native-web/src/exports/Image/index.js:163]: I noticed you updated this in #13 to:

```suggestion
  if (a.uri !== b.uri) return true;

  const headersA = a.headers || {};
  const headersB = b.headers || {};

  return JSON.stringify(headersA) !== JSON.stringify(headersB);
```

Comment (Beamanator) [reply packages/react-native-web/src/exports/Image/index.js:163]: Aah this is because in #13, `headers` may or may not exist

**Review by Beamanator — DISMISSED** [review]
@kidroca thanks for much for keeping this PR around while also showing what #13 would look like - to be honest I am now thinking the `ImageLoader` changes look much simpler in this PR, and even I like how the logic for the `ImageWithHeaders` is separated out, so I'm actually go back on my previous thought, and I'd say this is good to merge 👍

**Comment by Beamanator** [comment]
@marcaaron I'll leave this unmerged for today in case you want to check out this one vs #13 - if you don't have any concerns (or if you're too busy today) I'm pretty confident we can move forward here so I'll merge tomorrow

**Comment by marcaaron** [comment]
Sounds good - I will take a look now! 

Also, just wanted to clear the air on [this comment](https://github.com/Expensify/react-native-web/pull/11#issuecomment-1385831859). Chatted 1:1 with Alex and it's clear we had different expectations about how this PR should be led and who should do the bulk of the reviewing. 

In retrospect, I should have clarified that stuff before leaving that comment. And it was wrong to imply that there were any bad intentions (definitely did not mean for it to come across that way). It also didn't need to be shared in public. Sorry Alex!

File: packages/react-native-web/src/exports/Image/index.js
Comment (marcaaron) [inline packages/react-native-web/src/exports/Image/index.js]:
```suggestion
    if (!hasSourceDiff(nextSource, request.current.source)) {
        return;
    }
```

File: packages/react-native-web/src/exports/Image/index.js
Comment (marcaaron) [inline packages/react-native-web/src/exports/Image/index.js]:
```suggestion
    if (onLoadStart) {
        onLoadStart();
    }
```

File: packages/react-native-web/src/exports/Image/index.js
Comment (marcaaron) [inline packages/react-native-web/src/exports/Image/index.js:378]: I think this line would be a good place to leave a comment that explains [this comment](https://github.com/Expensify/react-native-web/pull/11/files#r1084967401).

File: packages/react-native-web/src/exports/Image/index.js
Comment (marcaaron) [inline packages/react-native-web/src/exports/Image/index.js]: Not too sure what this comment means. Is there a different way to say this? 

I think it's something like - the `BaseImage` `onLoadStart` event is not exposed to the parent. We are only interested in when the source with headers starts loading and not when the `BaseImage` loading starts?

Comment (marcaaron) [inline packages/react-native-web/src/exports/Image/index.js]:
```suggestion

    // Omit `onLoadStart` because we trigger it in the current scope
```

File: packages/react-native-web/src/exports/Image/index.js
Comment (marcaaron) [inline packages/react-native-web/src/exports/Image/index.js:397]:
```suggestion

    // Until the current component resolves the request (using headers)
```

File: packages/react-native-web/src/modules/ImageLoader/index.js
Comment (marcaaron) [inline packages/react-native-web/src/modules/ImageLoader/index.js]:
```suggestion
    const abortController = new AbortController();
```

File: packages/react-native-web/src/modules/ImageLoader/index.js
Comment (marcaaron) [inline packages/react-native-web/src/modules/ImageLoader/index.js]:
```suggestion
      signal: abortController.signal
```

File: packages/react-native-web/src/modules/ImageLoader/index.js
Comment (marcaaron) [inline packages/react-native-web/src/modules/ImageLoader/index.js]:
```suggestion
        abortController.abort();
```

**Review by marcaaron — CHANGES_REQUESTED** [review]
Code looks great. Love this new direction. I only have a few style notes and requests to improve the comments in the code.

File: packages/react-native-web/src/exports/Image/index.js
Comment (Beamanator) [reply packages/react-native-web/src/exports/Image/index.js]: I think it's more like: `ImageWithHeaders` already calls `onLoadStart` when it starts loading the image, so we don't want the `BaseImage` to trigger that function a second time

Comment (kidroca) [reply packages/react-native-web/src/exports/Image/index.js]: Now that I look at it I see it's confusing - it's like Alex said - loading starts inside `ImageWithHeaders`, to prevent `BaseImage` to raise `onLoadStart` a second time we filter it out

**Comment by kidroca** [comment]
I'll apply the suggestions to the PR and post back in a minute

**Comment by kidroca** [comment]
@Beamanator @marcaaron 
I've applied the suggested changes and updated / added code comments 

I've also updated the PR in App https://github.com/Expensify/App/pull/13036 to use the latest changes made here
I think it's looking good

File: packages/react-native-web/src/exports/Image/index.js
Comment (Beamanator) [inline packages/react-native-web/src/exports/Image/index.js:353]: I'm a bit confused by this comment - isn't `ImageSource` just a type? And is 'source' in the comment 'props.source'?

I think you're referring to `if (props.source && props.source.headers) {` check which will render this, component down below, right?

File: packages/react-native-web/src/exports/Image/index.js
Comment (Beamanator) [inline packages/react-native-web/src/exports/Image/index.js]:
```suggestion
    // Store a ref for the current load request so we know what's the last loaded source,
    // and so we can cancel it if a different source is passed through props
```

**Review by Beamanator — CHANGES_REQUESTED** [review]
Two more tiny questions on comments

**Comment by Beamanator** [comment]
Oh and one more question / possible point of confusion:

`ImageWithHeaders` loads an image with headers....
- While it's loading, it passes `source: undefined` to `BaseImage` (since the `blobUri` hasn't been set yet and so `BaseImage` doesn't try to load any image yet)
- After it loads, it passes `source: { nextSource: { headers: {...} }, uri: blobUri}` to `BaseImage`
    - From here, `BaseImage` treats `blobUri` as a normal `uri` and loads it like a normal image (via `ImageLoader.load`), firing `onError` and / or `onLoadEnd` if that errors out, right?
    - So we need to handle errors / loading ending in both `BaseImage` and in `ImageWithHeaders` because `ImageWithHeaders` will technically have 2 types of loading stages, each of which "can" error out and then need to call `onLoadEnd`?

Just making sure my understanding is correct 😅

File: packages/react-native-web/src/exports/Image/index.js
Comment (marcaaron) [reply packages/react-native-web/src/exports/Image/index.js]: Ok cool I think we are all saying the same thing - `BaseImage` will not use an `onLoadStart` callback when we have headers.

File: packages/react-native-web/src/exports/Image/index.js
Comment (marcaaron) [inline packages/react-native-web/src/exports/Image/index.js:394]: This is great thanks 👍

**Review by marcaaron — DISMISSED** [review]

File: packages/react-native-web/src/exports/Image/index.js
Comment (kidroca) [reply packages/react-native-web/src/exports/Image/index.js:353]: > I think you're referring to `if (props.source && props.source.headers) {` check which will render this, component down below, right?

Yes, I'm referring to the prior check
I'm trying to explain that `source` would always be specific subset of the `Source` type which is `ImageSourde` - an object with headers

There's probably a `flow` way to declare this
like making `ImageWithHeaders` use different props, particularly a `source` prop of type `ImageSource` 

I can change that to `$FlowFixMe` if you want to

**Comment by kidroca** [comment]
@Beamanator 
> Oh and one more question / possible point of confusion:
> 
> `ImageWithHeaders` loads an image with headers....
> 
> * While it's loading, it passes `source: undefined` to `BaseImage` (since the `blobUri` hasn't been set yet and so `BaseImage` doesn't try to load any image yet)
> * After it loads, it passes `source: { nextSource: { headers: {...} }, uri: blobUri}` to `BaseImage`
>   
>   * From here, `BaseImage` treats `blobUri` as a normal `uri` and loads it like a normal image (via `ImageLoader.load`), firing `onError` and / or `onLoadEnd` if that errors out, right?
>   * So we need to handle errors / loading ending in both `BaseImage` and in `ImageWithHeaders` because `ImageWithHeaders` will technically have 2 types of loading stages, each of which "can" error out and then need to call `onLoadEnd`?
> 
> Just making sure my understanding is correct 😅

I think you've got it right

When we load image with headers the `fetch` request might fail - we call `onError` and `onLoadEnd` there 
Otherwise (if `fetch` succeeds) `onLoadEnd` would be called from `BaseImage` 
`BaseImage` might still raise `onError` if the `fetched` file is not an image 

We only need to take care of the `Promise` rejection in `ImageWithHeaders` and raise error / loadEnd, `BaseImage` would take care of the rest like it already does

File: packages/react-native-web/src/exports/Image/index.js
Comment (Beamanator) [reply packages/react-native-web/src/exports/Image/index.js:353]: Cool cool nah let's keep it 👍

**Review by Beamanator — APPROVED** [review]
Thanks for all the work on this @kidroca 👍 👍

**Comment by Beamanator** [comment]
Published in v0.18.11 👍

**Comment by roryabraham** [comment]
@Beamanator @kidroca can you please include a link to the upstream PR in the description to make it easier to track this change in the upstream?

**Comment by Beamanator** [comment]
@roryabraham sure thing - the upstream PR is: https://github.com/necolas/react-native-web/pull/2442 (I added to the OP too!)

**Pull request merged**
